### PR TITLE
shelley block CBOR serialization

### LINKED ISSUE #1078: serialize block
The PR should finish the remaining Shelley serialization, namely transaction witnesses and the complete block.

### PR DESCRIPTION
**Pull request by JaredCorduan**
This PR completes the CBOR serialization for blocks in shelley.

closes #1078

### CODE REVIEW

File: shelley/chain-and-ledger/executable-spec/src/BlockChain.hs [hunk 1]
```diff
@@ -244,6 +253,134 @@ data Block crypto
...
+    SegWitsAll ws ss ->
+      encodeListLen 3
+        <> toCBOR (4 :: Word8)
+        <> toCBOR ws
+        <> toCBOR ss
+
+instance Crypto crypto
+  => FromCBOR (SegWits crypto)
+ where
+  fromCBOR = decodeListLen >> decodeWord >>= \case
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/BlockChain.hs]: We should compare the value from `decodeListLen` to the size of the record we get here

File: shelley/chain-and-ledger/executable-spec/src/BlockChain.hs [hunk 2]
```diff
@@ -244,6 +253,134 @@ data Block crypto
...
+    2 -> do
+      a <- fromCBOR
+      b <- fromCBOR
+      pure $ SegWitsScript a b
+    3 -> SegWitsScripts <$> fromCBOR
+    4 -> do
+      a <- fromCBOR
+      b <- fromCBOR
+      pure $ SegWitsAll a b
+    k -> fail $ "not a valid key: " ++ show k
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/BlockChain.hs]: Let's use our `invalidKey` function here

File: shelley/chain-and-ledger/executable-spec/src/BlockChain.hs [hunk 3]
```diff
@@ -244,6 +253,134 @@ data Block crypto
...
+        <> toCBOR wits
+      where
+        bodies = CborSeq $ fmap _body txns
+        wits = CborSeq $ fmap txToSegWit txns
+
+instance Crypto crypto
+  => FromCBOR (Block crypto)
+ where
+  fromCBOR = do
+    _ <- decodeListLen
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/BlockChain.hs]: Maybe `enforceSize 3` here?

File: shelley/chain-and-ledger/executable-spec/src/BlockChain.hs (hunk 1 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/BlockChain.hs]: yes indeed, thanks!

**Pull request merged**